### PR TITLE
Change CurlHandler to use IOExceptions for Stream failures

### PR DESCRIPTION
**Pull request by stephentoub**
As with WinHttpHandler, CurlHandler wraps exceptions in HttpRequestException.  But WinHttpHandler switches to using IOException when sending exceptions out via Stream.Read/WriteAsync.  This change does the same for CurlHandler.

This also adds a test for #6231, which was correctly failing on Unix but with the wrong exception type (which is now the correct type with the previous commit).

cc: @davidsh, @joelverhagen, @ericeil, @kapilash

### CODE REVIEW

File: src/System.Net.Http/tests/FunctionalTests/ResponseStreamTest.cs [hunk 1]
```diff
@@ -90,6 +91,54 @@ await client.GetAsync(HttpTestServers.RemoteEchoServer, HttpCompletionOption.Res
...
+                    while (!string.IsNullOrEmpty(line = reader.ReadLine())) ;
+
+                    // Write response, with a potentially invalid content length
+                    using (var writer = new StreamWriter(stream, Encoding.ASCII))
+                    {
+                        string content = "This is some response content.";
+                        writer.Write("HTTP/1.1 200 OK\r\n");
+                        writer.Write($"Date: {DateTimeOffset.UtcNow:R}\r\n");
+                        writer.Write("Content-Type: text/plain\r\n");
+                        writer.Write($"Content-Length: {content.Length + 42}\r\n"); // incorrect length
```
Comment (bartonjs) [inline src/System.Net.Http/tests/FunctionalTests/ResponseStreamTest.cs]: I didn't think we were able to use this construct yet?  The TFS mirrored builds are still C# 5, I think?

**Comment by bartonjs** [comment]
As long as it's not going to break the mirror, structurally LGTM.

File: src/System.Net.Http/tests/FunctionalTests/ResponseStreamTest.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/ResponseStreamTest.cs]: Fixed with #6181.  Hence the very large https://github.com/dotnet/corefx/pull/6209 :smile:

Comment (bartonjs) [inline src/System.Net.Http/tests/FunctionalTests/ResponseStreamTest.cs]: Ah, good to know.

**Comment by davidsh** [comment]
Since this is part of unblocking an important scenario I am ok with it for now. I will try to add new capability to Azure within the next few days for this. In particular, we really need to test a few different "incomplete message read" scenarios.  'Content-Length' having length than expected is one scenario.  But there is also 'Transfer-Encoding: chunked' scenario where the connection is closed prior to the server sending the zero-length chunk.  Finally, there is another response-test scenario where payload is sent without any Content-Length or Transfer-Encoding semantics. In that case, there is no "expected" playload length.

File: src/System.Net.Http/tests/FunctionalTests/ResponseStreamTest.cs [hunk 2]
```diff
@@ -90,6 +91,54 @@ await client.GetAsync(HttpTestServers.RemoteEchoServer, HttpCompletionOption.Res
             }
         }
 
+        [ActiveIssue(6231, PlatformID.Windows)]
+        [Fact]
+        public async Task ReadAsStreamAsync_InvalidContentLengthResponse_ThrowsIOException()
```
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/ResponseStreamTest.cs]: I would change "InvalidContentLengthResponse" to "IncompleteContentLengthResponse"

Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/ResponseStreamTest.cs]: Fixed.

**Comment by stephentoub** [comment]
Test Innerloop OSX Release Build and Test please
Test Innerloop OSX Debug Build and Test please
Test Innerloop CentOS7.1 Release Build and Test please

**Pull request merged**